[FLINK-16989][table] Support ScanTableSource in blink planner

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Support the `ScanTableSource` interface in planner. We don't support change mode `[I, UA, D]` (`UPDATE_AFTER` without `UPDATE_BEFORE`) in first version. May support it in the future. 

## Brief change log

- We rename the existing `TableSourceTable` into `LegacyTableSourceTable` and `TableSourceScan` into `LegacyTableSourceScan`, and use `TableSourceTable` and `TableSourceScan` for the new `DynamicTableSource`.
  - We don't want to use `LogicalDynamicTableSourceScan` because the `Dynamic` is verbose and misleading. 
  - Besides, `LegacyTableSourceScan` will be dropped in the near future, we can easily remove all `Legacy` classes at that time. 
- Implements logical node and physical nodes for `ScanTableSource` and integrate with changelog program.

## Verifying this change

- Introudce a `TestValuesTableFactory` for testing
- Use the `values` source in `TableScanTest` and add validation tests in it. 
- Add `DynamicTableSourceITCase` for integration tests 
- Add `testChangelog` and `testAggregateOnChangelogSource` for integration tests for changelog source.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not applicable

danny0405 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala:
```diff
@@ -226,4 +251,113 @@ class CatalogSourceTable[T](
+        }
+      case _ =>
+        unsupportedAbilities.foreach { ability =>
+          if (tableSource.getClass.isAssignableFrom(ability)) {
```
Does the `ScanTableSource ` support all these features now ?

wuchong replied:
No. We don't support any of the ability interface. That's why we should throw exception to idnicate users they are not supported yet. Once one of the ability interface is supported, the developer should remove then entry from `unsupportedAbilities`.

danny0405 replied:
Did i miss something, the `ScanTableSource` code would never go here. So the check never triggers.

wuchong replied:
I moved this validation to the beginning of this method. And added a test for it.

danny0405 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala:
```diff
@@ -226,4 +251,113 @@ class CatalogSourceTable[T](
+              s"${ability.getSimpleName} ability is not supported.")
+          }
+        }
+    }
```
How about we move to a new CatalogSourceTable and keep the old one as `LegacyCatalogSourceTable` ? The code now is a mess.

danny0405 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala:
```diff
@@ -226,4 +251,113 @@ class CatalogSourceTable[T](
+        // success, then we will use the legacy factories
+        true
+      } catch {
+        case _: Throwable =>
```
Should we use the connector.property-version to identify if the connector is new or not ?

wuchong replied:
I though about this. However, it doesn't work. Because, the `connector.property-version` is attached by the framework. So no matter the the connector is old or new, they have the same `connector.property-version`.

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTableSourceScan.scala:
```diff
@@ -92,177 +73,21 @@ class StreamExecTableSourceScan(
+  override protected def translateToPlanInternal(
+      planner: StreamPlanner): Transformation[RowData] = {
+    createSourceTransformation(planner.getExecEnv, getRelDetailedDescription)
   }
```
Yes. This is on purpose. We should prefer to use `getRelDetailedDescription` which have more detailed and formated information about this node. Besides, the `DynamicTableSource#asSummaryString` is not the same meaning with `TableSource#explainSource`. `asSummaryString` is just used for logging (such as `Kafka_0.9`), it doesn't contain detailed information of this node.

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/trait/UpdateKindTrait.scala:
```diff
@@ -95,4 +98,17 @@ object UpdateKindTrait {
+      case (true, true) => BEFORE_AND_AFTER
+      case (false, true) => ONLY_UPDATE_AFTER
+      case _ => NONE
+    }
```
OK. I will throw exception for `true, false`, but keep `NONE` for `false, false`.

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/logical/FlinkLogicalTableSourceScan.scala:
```diff
@@ -76,7 +73,7 @@ class FlinkLogicalTableSourceScan(
     super.explainTerms(pw)
-      .item("fields", tableSource.getTableSchema.getFieldNames.mkString(", "))
+      .item("fields", getRowType.getFieldNames.mkString(", "))
   }
```
We don't have `getTableSchema` method for `DynamicTableSource`. If we support projection push down, the `getRowType` should be the projected fields.

danny0405 commented on flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/DynamicTableSourceITCase.scala:
```diff
@@ -0,0 +1,255 @@
+    val dataId = TestValuesTableFactory.registerData(data)
+
+    // TODO: [FLINK-17428] support nested project for TestValuesTableSource
+    val ddl =
```
Also add tests for computed column and watermark definition ?

wuchong replied:
Sure.

wuchong commented:
Hi @danny0405 , thanks for reviewing. I have addressed the comments. Some major changes:

1. I updated `TimeAttributeITCase` to use the new factories as sources which covers watermark and computed column.
2. Keep the implementation of the original `CatalogSourceTable` and rename to `LegacyCatalogSourceTable`.
3. Rename test classes `DynamicTableSourceITCase` to `ScanTableSourceITCase`.

danny0405 commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/FlinkCalciteCatalogReader.java:
```diff
@@ -188,11 +192,46 @@ private static FlinkPreparingTableBase convertCatalogTable(
+			CatalogTable catalogTable,
+			CatalogSchemaTable schemaTable) {
+		if (catalogTable.getOptions().containsKey(ConnectorDescriptorValidator.CONNECTOR_TYPE)) {
+			return true;
```
I think we should consider the case sensitivity here.

wuchong replied:
Why? I checked `TableFactoryService`, we use lower-case as the option key. And `CONNECTOR_TYPE` is lower case.

danny0405 replied:
Because `catalogTable.getOptions()` returns the original key.

danny0405 commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala:
```diff
@@ -226,4 +176,89 @@ class CatalogSourceTable[T](
+            case (true, true) =>
+              // UPDATE_BEFORE and UPDATE_AFTER, pass
+            case (false, true) =>
+              // only UPDATE_AFTER
```
I saw similar logic in `FlinkChangelogModeInferenceProgram`, maybe we should have a `ChangelogMode` util for such a validation ~

wuchong replied:
Do you mean in `UpdateKindTrait`? It's not similar. There is not much code can be shared between them.

wuchong commented:
Azure build is passed in my private pipeline: https://dev.azure.com/imjark/Flink/_build/results?buildId=58&view=results

Merging...

wuchong commented:
The just merged #11905 conflicts with this PR. Wait another round of build...
